src/ripple/shamap/README.md

```diff
@@ -26,164 +37,307 @@ There are two different ways of building and using a SHAMap:
+SHAMap can be handled using raw pointers (if you're careful).
 
 One consequence of this design is that an immutable SHAMap can never be
-"trimmed".  There is no way to identify unnecessary nodes in an immutable
-SHAMap that could be removed.  Once a node has been brought into the
-in-memory SHAMap, that node stays in memory for the life of the SHAMap.
+"trimmed".  There is no way to identify unnecessary nodes in an immutable SHAMap
+that could be removed.  Once a node has been brought into the in-memory SHAMap,
+that node stays in memory for the life of the SHAMap.
 
-Most SHAMaps are immutable, in the sense that they don't modify or remove
-their contained nodes.
+Most SHAMaps are immutable, in the sense that they don't modify or remove their
+contained nodes.
 
 An example where a mutable SHAMap is required is when we want to apply
-transactions to the last closed ledger.  To do so we'd make a mutable
-snapshot of the state tree and then start applying transactions to it.
-Because the snapshot is mutable, changes to nodes in the snapshot will not
-affect nodes in other SHAMAps.
+transactions to the last closed ledger.  To do so we'd make a mutable snapshot
+of the state tree and then start applying transactions to it. Because the
+snapshot is mutable, changes to nodes in the snapshot will not affect nodes in
+other SHAMAps.
+
+An example using a immutable ledger would be when there's an open ledger and
+some piece of code wishes to query the state of the ledger.  In this case we
+don't wish to change the state of the SHAMap, so we'd use an immutable snapshot.
+
+## Sequence numbers ##
+
+Both SHAMap's and their nodes carry a sequence number.  This is simply an
+unsigned number that indicates ownership or membership, or a non-membership.
+
+SHAMap's sequence numbers normally start out as 1.  However when a snap-shot of
+a SHAMap is made, the copy's sequence number is 1 greater than the original.
 
-An example using a immutable ledger would be when there's an open ledger
-and some piece of code wishes to query the state of the ledger.  In this
-case we don't wish to change the state of the SHAMap, so we'd use an
-immutable snapshot.
+The nodes of a SHAMap have their own copy of a sequence number.  If the SHAMap
+is not immutable, meaning it can change, then all of its nodes must have the
+same sequence number as the SHAMap itself.  This enforces an invariant that none
+of the nodes are shared with other SHAMaps.
 
+When a SHAMap gets a copy of a node from elsewhere, and must have a private
+copy, it clones the node and sets the nodes sequence number equal to the SHAMap
+sequence number.
+
+When a SHAMap decides that it is safe to share a node of its own, it sets the
+node's sequence number to 0 (a SHAMap never has a sequence number of 0).
+
+Note that other objects in rippled also have sequence numbers (e.g. ledgers).
+The SHAMap and node sequence numbers should not be confused with these other
+sequence numbers (no relation).
```

Comment:
Since we touched on sequence numbers, can we also talk about concurrency? How does shamap handle concurrency?

Reply:
This is discussed in the Canonicalize section.  If you have a better way to describe this, I'm open to suggestions.